### ISSUE TITLE
Inpainting models barely apply any change even at denoise 1

### ISSUE THREAD

**Issue by ramyma** [comfyanonymous/ComfyUI#1186]
When using an inpainting model there's almost no change happening to the masked region, even with a denoise value of 1.
However, when using the same flow with a regular model it seems to work.

With inpainting model:
![Screenshot from 2023-08-11 03-51-21](https://github.com/comfyanonymous/ComfyUI/assets/14105815/8f6f24ef-d326-47b6-8f3f-da62a0f66555)

With regular model:
![Screenshot from 2023-08-11 03-56-49](https://github.com/comfyanonymous/ComfyUI/assets/14105815/55b55283-2618-41f2-a469-ff795be0dc60)

**Comment by comfyanonymous** [comment]
Inpaint models need the "VAE Encode for inpaint" node or else it does that.

**Comment by ramyma** [comment]
But with "VAE Encode for inpaint" it's not usable with low denoise values; as it adds a grey fill to the masked region

**Comment by comfyanonymous** [comment]
That grey fill is what the inpaint model expects.

**Comment by ramyma** [comment]
This is how it looks with denoise 0.5:
![Screenshot from 2023-08-11 04-15-43](https://github.com/comfyanonymous/ComfyUI/assets/14105815/1ec8b723-bea6-4414-b2db-3f7a1c48dd03)

**Comment by ramyma** [comment]
Is there a way to replicate A1111's inpaint with an inpainting model and fill method "original?

**Comment by ltdrdata** [comment]
> Is there a way to replicate A1111's inpaint with an inpainting model and fill method "original? 

You have to use SetLatentNoise instead of VAEEncode (for inpaint)

**Comment by ramyma** [comment]
> > Is there a way to replicate A1111's inpaint with an inpainting model and fill method "original?
> 
> You have to use SetLatentNoise instead of VAEEncode (for inpaint)

@ltdrdata  VAEEncode should be used with inpainting models accourding to @comfyanonymous:
https://github.com/comfyanonymous/ComfyUI/issues/1186#issuecomment-1674105576

**Comment by pennyvc** [comment]
If you want to use any of the original image, im fairly sure you need to make use of the setlatentnoise method and normal model rather than inpaint model.

**Comment by ramyma** [comment]
I've already tried the proposed workflows, however, the reason I opened this issue is that there is a missing functionality when using **inpainting models**; having a denoise value lower than 1 while using the original pixel values of the input image, this is one of the main benefits of using an inpainting model.
Here's an example from A1111:
![Screenshot from 2023-08-11 13-32-18](https://github.com/comfyanonymous/ComfyUI/assets/14105815/814ed288-e6e5-4067-9dc3-ccc2d3d27d89)

As you can see the consistency is much better than using a denoise value of 1 with "VAE Encode for inpaint" and an inpainting model.

**Comment by Lalimec** [comment]
Did you find a solution for this, that damn gray also affects the color of inpainting if the area is large.

**Comment by ramyma** [comment]
> Did you find a solution for this, that damn gray also affects the color of inpainting if the area is large.

No, I was hoping I would get an answer here. I believe it's a issue/missing feature.

**Comment by Lalimec** [comment]
did you try this https://github.com/comfyanonymous/ComfyUI/issues/668

**Comment by ramyma** [comment]
> did you try this #668
Yes, but it doesn't seem to have a noticeable effect with inpainting models

**Comment by sinelka** [comment]
I also had the same problem. Even if a month has passed, I will leave my observations here, maybe it will be useful to someone. The inpaint model really doesn't work the same way as in A1111. It is necessary to use VAE Encode (for inpainting) and select the mask exactly along the edges of the object. In the first example (Denoise Strength 0.71), I selected only the lips, and the model repainted them green, almost leaving a slight smile of the original image. In the second example (Denoise Strength 0.8) I masked the entire mouth area - and the model drew green lips, but lost the original concept. Set Latent Noise Mask does not work for the Inpaint model.
![2023-09-14_19-36-06](https://github.com/comfyanonymous/ComfyUI/assets/145046934/ec9e4d36-a137-470e-83cf-a0c95fa222ce)
![2023-09-14_19-45-23](https://github.com/comfyanonymous/ComfyUI/assets/145046934/b5b13ec1-5464-400e-97ab-fb0779184e1c)
Sorry for a little mess on a screen :)

**Comment by ChieVFX** [comment]
Definitely agree that it's an issue and I hope ot gets resolved. Right now, inpaintng in ComfyUI is deeply inferior to A1111, which is letdown..

I'll reiterate:
Using "Set Latent Noise Mask" allow you to lower denoising value and get profit from information already on the image(e.g. you sketched something yourself), but when using Inpainting models, even denoising of 1 will give you an image pretty much identical to the original..

When not using Inpainting model - you can use "Set Latent Noise Mask" approach, however, using A1111 with lower denoise value and Inpainting model give much much better results. So, there is a lot of value of allowing us to use Inpainting model with "Set Latent Noise Mask".

The only way to use Inpainting model in ComfyUI right now is to use "VAE Encode (for inpainting)", however, this only works correctly with the denoising value of 1. And that means we can not use underlying image(e.g. sketch stuff ourselves).

---
What I expect: "Set Latent Noise Mask" works fine with inpainting models, changing image according to denoising value, same as A1111.
Or
"VAE encode (for inpainting)" with a lower denoising value should use underlying image instead of grey, when using lower denoising value.

What happens: "Set Latent Noise Mask" with inpainting model changes almost nothing with any denoising value(including 1), while "VAE encode (for inpainting)" only works correctly with denoising value of 1, hence making it imposible to use underlying image efficiently or at all.

**Comment by ansonkao** [comment]
Does anyone here understand both Comfy and A1111 codebases well enough to understand what they are doing differently to cause Comfy's inpainting behaviour to be so different? 

I spent a couple hours digging into how the VAEEncodeForInpaint, SetLatentNoiseMask, and KSampler nodes work, hoping I could submit a patch PR. So far I can see that the denoising/scheduling/masking logic appear to be equivalent, but I can't isolate how A1111 handles the different masking options differently yet. It's gonna take a lot more time to grok without help.

@comfyanonymous any chance you could point us in the right direction?

**Comment by ansonkao** [comment]
I've identified a key difference!

In Automatic1111 webui, time elapsed during img2img is directly proportional to denoising strength, i.e.:
- For denoising = 1.0, generation takes X seconds
- For denoising = 0.5, generation takes 0.5X seconds
- For denoising = 0.01, generation takes 0.01X seconds

In ComfyUI, time elapsed is constant regardless of the denoising strength

Now I know what to look for in the codebase! Stay tuned

**Comment by thenickdude** [comment]
@ansonkao, yes, this is because denoise controls what is exposed at "start_at_step" parameter of the Advanced Sampler. Basically for denoise 0.5, the sampler skips the first 50% of the steps and uses the input latent image in place of those.

**Comment by jn-jairo** [comment]
> * For denoising = 1.0, generation takes X seconds
> * For denoising = 0.5, generation takes 0.5X seconds
> * For denoising = 0.01, generation takes 0.01X seconds

The A1111 uses less steps proportional to the denoising, that is why it is faster with less denoising, there is a configuration in A1111 that disables it, to have the same in ComfyUI just use less steps.

**Comment by Gloynus** [comment]
> то же самое в ComfyUI, просто используйте меньше шагов

It won't work. At 1-2 steps it will be low quality, but at 3 you will already see the original.

**Comment by ntrouve-onera** [comment]
Bump, no update on this issue? From my understanding Inpainting models use a mask as an extra input, the one fed to the inpainting model is wrong for some reason when using Masked Latent Node. Resulting in the inpainting model using the fill data as a reference to inpaint (and as the fill is actually decent see no reason to change anything).

This is a huge limitation especially when trying to use Lama as a preprocessor to get a rough fill before using the inpaint model.

**Comment by JoeNavark** [comment]
I haven't fully tested it yet but it seems I can get around this issue by using a normal vaeencode on the original image along with the vaeencode for inpainting and doing a latentblend.

**Comment by ntrouve-onera** [comment]
@JoeNavark and with 50% blend you get roughly what you could have expected from 0.5 denoise? (more or less?).

**Comment by ntrouve-onera** [comment]
![image](https://github.com/comfyanonymous/ComfyUI/assets/64471238/e76a5020-4f05-4274-b7cf-ab62ccfa3d43)

This does somewhat work, you need to add an extra latent mask or it just inpaint everything. Still some feathering/edge issues to solve but it appears to be doing what we want.

**Comment by JoeNavark** [comment]
That's weird, I didn't have to add the second mask as long as the masked encode went to samples1.

**Comment by ntrouve-onera** [comment]
I may have switched samples1 and 2 before doing the screenshot to clean up so it might be because of that.

**Comment by xucj98** [comment]
I solved this problem by modifying some code. The Inpainting model requires receiving "noisy images" and "masked images" as inputs. WEBUI and ComfyUI use different processing methods, as shown in the following figure.

![WcfuBkqSMi](https://github.com/comfyanonymous/ComfyUI/assets/28194163/e3d9f1e0-3b14-4441-9f76-e7d5c946e1ca)

To solve this problem, the code in four places of the three files needs to be modified.
1. nodes.VAEEncodForInpaint.encode() , need to output "latent_image" and "masked_ latent_ image"
2. nodes.common_ksampler(), need to add "masked_latent_image" to positive and negative condition
3. comfy.sampler.sample(), do model.process_latent_in() on the "masked_latent_image"
4. comfy.model_base.BaseModel.extra_conds(), use "masked_latent_image" instead of "latent_image" as element of "cond_concat"  

If the project manager agrees to this plan, I would be happy to submit a PR to fix this issue.

**Comment by ltdrdata** [comment]
> I solved this problem by modifying some code. The Inpainting model requires receiving "noisy images" and "masked images" as inputs. WEBUI and ComfyUI use different processing methods, as shown in the following figure.
> 
> ![workflow](https://github.com/comfyanonymous/ComfyUI/assets/28194163/961b9131-fae0-406d-99fa-671b9f5f4dba)
> 
> 
> To solve this problem, the code in four places of the three files needs to be modified.
> 1. nodes.VAEEncodForInpaint.encode() , need to output "latent_image" and "masked_ latent_ image"
> 2. nodes.common_ksampler(), need to add "masked_latent_image" to positive and negative condition
> 3. comfy.sampler.sample(), do model.process_latent_in() on the "masked_latent_image"
> 4. comfy.model_base.BaseModel.extra_conds(), use "masked_latent_image" instead of "latent_image" as element of "cond_concat"  
> 
> If the project manager agrees to this plan, I would be happy to submit a PR to fix this issue.
> 
> 
> 
> 

Finally, an analysis has been conducted to identify the implementation differences. Good job!

**Comment by ntrouve-onera** [comment]
Thanks a lot, supporting the idea of the PR. Indeed the workaround is still not as good as a true partial denoise because the latent blend appears to mess up with noise offset resulting in some tint missmatch.

**Comment by xucj98** [comment]
I submitted a PR [#2501](https://github.com/comfyanonymous/ComfyUI/pull/2501)
